ENT-215/ENT-217: Set up data transformation framework/do transformations for SAP

haikuginger opened the pull request:
**Description:** 

This PR adds a framework for course metadata transformation/transmission to integrated channels.

**JIRA:** [ENT-215](https://openedx.atlassian.net/browse/ENT-215)

**Dependencies:** None

**Merge deadline:** 13 March 2017

**Installation instructions:** Requires course discovery devstack populated with course metadata

**Testing instructions:**

1. Create an `EnterpriseCustomer` with a catalog attached, and with courses present in the catalog.
2. Create an `SAPSuccessFactorsEnterpriseCustomerConfiguration` linked to the EnterpriseCustomer
3. In a devstack, run the SAP management command, and verify that the logging output shows the JSON structure that are expected for SAP. Be sure to use a username with access to the catalog in use:
`./manage.py lms transmit_courseware_data --catalog_user staff`

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

edx-webhook commented:
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=63"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

haikuginger commented:
@brittneyexline @mattdrayer @pomegranited, this is ready for a first pass. Note that this is based on @pomegranited's current PR and currently is lugging all of her commits around.

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -64,6 +64,30 @@ def get_all_catalogs(self):
         """
         return self._load_data('catalogs', default=[])
 
+    def get_single_catalog_courses(self, catalog_id):
```
why `get_single_catalog_courses` instead of `get_catalog_courses`?  Is the latter already in-use?

haikuginger replied:
Will change.

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -64,6 +64,30 @@ def get_all_catalogs(self):
+        Returns:
+            dict: Courses of the catalog in question
+        """
+        return self._load_data('catalogs/{catalog_id}/courses'.format(catalog_id=catalog_id))
```
Do we need to protect against an invalid `catalog_id` value?

haikuginger replied:
We can add some protection, but it shouldn't be necessary - the upstream catalog utils will give us an empty list if the catalog ID isn't valid.

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -64,6 +64,30 @@ def get_all_catalogs(self):
+        """
+        return self._load_data('catalogs/{catalog_id}/courses'.format(catalog_id=catalog_id))
+
+    def get_single_course_details(self, course_key):
```
Here as well -- why `get_single_course_details` instead of `get_course_datails'?

haikuginger replied:
Agreed.

mattdrayer commented on enterprise/course_catalog_api.py:
```diff
@@ -64,6 +64,30 @@ def get_all_catalogs(self):
+        Returns:
+            dict: Details of the course in question.
+        """
+        return self._load_data('courses', resource_id=course_key, many=False)
```
Are we concerned at all about invalid `course_key` values?

haikuginger replied:
@mattdrayer: No. As above, the edX API utils will give us an empty result if the request fails.

@pomegranited: https://github.com/edx/edx-platform/blob/master/openedx/core/lib/edx_api_utils.py#L17-L18

pomegranited replied:
Lol whoops..  I was on the wrong fork!  Sorry about that @haikuginger .

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+LOGGER = getLogger(__name__)
+
+
+def updated_course_run(course_details, course_run_details):
```
This function is named like a property, but it's performing an action -- maybe better to name it `update_course_run_details`?

pomegranited replied:
Or `get_complete_course_run_details` ?  Or `merge_course_run_details`?

haikuginger replied:
I like `get_complete_course_run_details`; will go with that.

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+
+def get_course_list(user, enterprise_customer):
+    """
+    Determine which catalog belongs to a given Enterprise Customer and list the course runs it contains.
```
Context-wise it doesn't matter to the caller if we are using a catalog to obtain the enterprise customer's list of course runs.  I'd drop that bit from the docstring.  Also you might consider naming this function `get_course_run_list` or `get_course_runs`.  If knowing that the catalog _is_ important to the caller, then we might name the function `get_catalog_course_runs`.

haikuginger replied:
Good call.

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+    catalog_id = enterprise_customer.catalog
+
+    if catalog_id is None:
+        LOGGER.info('No catalog attached to %s', enterprise_customer)
```
Is it okay to fail silently here and return None?

pomegranited replied:
I think it's best to filter out any `enterprise_customer` objects with no `catalog` way back in the management command, and just raise here.

Or better yet, take a `catalog_id` parameter, instead of `enterprise_customer`.

haikuginger replied:
@mattdrayer, @brittneyexline: Since this is a generator, not having an explicit return value means that the caller gets an empty iterable back. I could make it more explicit by raising `StopIteration` if you'd prefer.

@pomegranited, that's a good call; it's a database restriction that we can make pretty easily.

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+        self.enterprise_customer = plugin_configuration.enterprise_customer
+        self.plugin_configuration = plugin_configuration
+
+    def get_parsed(self):
```
`get_parsed"...course_runs?  What are we actually doing when we say we are "parsing"?

haikuginger replied:
Inline with below, `get_transformed_course_runs`?

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+        results = (self.process_course_run(run) for run in get_course_list(self.user, self.enterprise_customer))
+        return results
+
+    def process_course_run(self, course_run_details):
```
I think I'd prefer something along the lines of `transform_course_run_details`

haikuginger replied:
I like it!

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+            output[key] = transform(course_run_details) if transform is not None else course_run_details.get(key)
+        return output
+
+    def send_all(self):
```
`send_all` what?  courses?  Is this more like `transmit_course_metadata`?

haikuginger replied:
`transmit_all_course_details`?

mattdrayer commented on integrated_channels/integrated_channel/course_metadata.py:
```diff
@@ -0,0 +1,127 @@
+        for course in self.get_parsed():
+            self.send(course)
+
+    def send(self, course):
```
As with the other function names it would be great if we could be more specific about what is going on -- eg `transmit_course_metadata`

brittneyexline replied:
I agree - especially to have some consistency with the learner data version. Is this where you would call the transmitter class I am creating?

haikuginger replied:
As above, will rename to be more specific.

@brittneyexline, this isn't exactly where you'll tap in - rather, you'll do that in the SAP-specific subclass of this class. But within this method, yes.

mattdrayer commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,86 @@
+    return int(time_since_epoch.total_seconds() * 1000)
+
+
+def in_interval(start, end):
```
How about `is_in_interval`

haikuginger replied:
I'll do you one better and make it `current_time_is_in_interval`!

mattdrayer commented on test_utils/fake_catalog_api.py:
```diff
@@ -235,6 +235,261 @@
+        "key": "edX+DemoX",
+        "uuid": "cf8f5cce-1370-46aa-8162-31fdff55dc7e",
+        "title": "Fancy Course",
+        "course_runs": [
```
Just a note that `track_selection_url` will be added to the course_runs dicts at some point

brittneyexline commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+            except EnterpriseCustomer.DoesNotExist:
+                raise CommandError('Enterprise customer {} not found, or not active'.format(enterprise_customer_uuid))
+            else:
+                enterprise_customers.append(enterprise_customer)
```
Is there a reason this can't be done in the try block? I'm not sure I understand what the if condition is for this else...

haikuginger replied:
For Python, you can do a `try-except-else` block; the try will catch errors raised inside that block, and the else will allow processing if no errors are raised. It lets us be specific about the particular lines of code we want to protect.

brittneyexline commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+            else:
+                enterprise_customers.append(enterprise_customer)
+        else:
+            enterprise_customer = None
```
If we only depend on enterprise_customers below, do we need this variable?

haikuginger replied:
In fact we don't. Good catch; residue of an earlier version.

brittneyexline commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+        enterprise_customer: The EnterpriseCustomer whose integrated channels we need to send.
+    """
+    user = User.objects.get(username=username)
+    enterprise_customer = EnterpriseCustomer.objects.get(uuid=enterprise_customer_uuid)
```
Could you pass the model in from the caller since you've already got the customers there, to avoid reading the object again?

haikuginger replied:
Celery requires that we be able to serialize the arguments we transmit to the task; strings can be JSON-serialized, but models cannot.

brittneyexline commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,86 @@
+                'launchURL': x['marketing_url'] or '',
+                'contentTitle': 'Course Description',
+                'contentID': x['key'],
+                'launchType': 2,
```
We will need to change this to 3

haikuginger replied:
Good to know! I tried to find the appropriate value in the doc, but couldn't find any details.

brittneyexline commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,86 @@
+        'content': lambda x: [
+            {
+                'providerID': SAPSuccessFactorsGlobalConfiguration.current().provider_id,
+                'launchURL': x['marketing_url'] or '',
```
I think @mattdrayer noted in some Jira ticket that we'll likely be sending the track selection url instead, but I don't think you need to change this just yet, it seems like it could be easily updated later. We also might have some other updates to this transform once we get some feedback from Griff.

haikuginger replied:
Sounds good. The whole point of this structure is to make it easy to write/change transforms for individual fields.

mattdrayer commented on integrated_channels/sap_success_factors/models.py:
```diff
@@ -113,6 +166,45 @@ def __repr__(self):
+            contactHours=empty_value,
+            cpeHours=empty_value,
+            instructorName=empty_value,
+            comments=empty_value,
```
@brittneyexline we should confirm with @griffresch on the empty fields we are sending here -- for example we may want to include some additional information in the 'comments' section.

pomegranited commented:
Also note: `Enterprise SAP SuccessFactors` is available in the LMS Admin since https://github.com/edx/edx-platform/pull/14563 merged (yay!) e.g.,

http://localhost:8000/admin/sap_success_factors/sapsuccessfactorsenterprisecustomerconfiguration/

haikuginger commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+except ImportError:
+    def celery_task(func):
+        """Use a no-op decorator if djcelery is not available."""
+        return func
```
Nice, thanks @pomegranited!

haikuginger commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+
+        # Find the correct user with which to access the catalog API
+        manual_catalog_username = options['catalog_user']
+        settings_catalog_username = getattr(settings, 'EDXAPP_ENTERPRISE_CATALOG_API_USERNAME', None)
```
This is a new setting, but it looks like (based on conversations with EdX people) we're going to want to switch to a configuration model.

haikuginger replied:
After some talk, we're removing this altogether; the command will just require a username passed as an argument.

haikuginger commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,133 @@
+    @staticmethod
+    def validate_user(username):
+        if not username:
+            raise CommandError('A username must be specified manually or in settings to access the course catalog API.')
```
Can do!

haikuginger replied:
So, with the above (removing the ability to set a global worker username), we can just flag this arg as required. Problem solved!

haikuginger commented:
Comments addressed; ready for another pass @pomegranited, @mattdrayer @brittneyexline.

pomegranited commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,96 @@
+        'status': lambda x: 'ACTIVE' if x['availability'] == 'Current' else 'INACTIVE',
+        'title': lambda x: [
+            {
+                'locale': 'English',
```
Is it ok to hard-code this locale?  I would have thought we'd have to [pull the locale out of Django somehow](http://stackoverflow.com/a/28048237/4302112) and map it to SAP's locale string.

Though it's probably this is out of scope for this first pass..  Maybe just make sure there's a follow-up ticket somewhere?

pomegranited commented on integrated_channels/integrated_channel/management/commands/transmit_courseware_data.py:
```diff
@@ -0,0 +1,81 @@
+            required=True,
+            metavar='ENTERPRISE_CATALOG_API_USERNAME',
+            help='Use this user to access the Course Catalog API. Omit this option to use the'
+            ' username specified in the EDXAPP_ENTERPRISE_CATALOG_API_USERNAME setting.'
```
nit: please update this comment to remove the "Omit this option.." part, since it's now required.

pomegranited commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,96 @@
+                    'sap_success_factors',
+                    'SAPSuccessFactorsGlobalConfiguration'
+                ).current().provider_id,
+                'launchURL': x['marketing_url'] or '',
```
Since @mattdrayer [said there's going to be a track selection URL eventually](https://github.com/edx/edx-enterprise/pull/63/files#r105011522), is it worth having here for when it does happen?  e.g.

```
 'launchURL': x['track_selection_url'] or x['marketing_url'] or '',
```

pomegranited commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,96 @@
+            },
+        ],
+        'thumbnailURI': lambda x: (x['image']['src'] or ''),
+        'content': lambda x: [
```
My understanding from the SAP docs was that this was supposed to be a list of sections in the course, with details about each section.

However, since this data isn't available from the Course Catalog, it's probably out of scope for this task.  It is available from the [Course Blocks API](http://edx.readthedocs.io/projects/edx-platform-api/en/latest/courses/blocks.html#example-response-showing-a-list-of-course-blocks-in-a-specified-course) though, if we need to have a follow-up task?

brittneyexline replied:
I think it makes sense for us to only send one content item representing the course as a whole from a logistical standpoint. It probably is worth exploring how sending multiple content items will behave and render in SAP's software and then decide if we want to take advantage of it.

pomegranited commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,96 @@
+    """
+    parsed_datetime = parse_lms_api_datetime(datestamp)
+    time_since_epoch = parsed_datetime - UNIX_EPOCH
+    return int(time_since_epoch.total_seconds() * 1000)
```
I thought the Unix epoch timestamps are usually expressed in seconds, not milliseconds?

If not, then [my PR needs fixing too](https://github.com/edx/edx-enterprise/pull/59/files#diff-b2d1ad7443d2344126a60d0213e2d192R150), which means we need a common method for this.

brittneyexline replied:
I am pretty sure it should be seconds.

pomegranited replied:
@brittneyexline Any update on this?  Should it be seconds or milliseconds?

I can fix this as part of https://github.com/edx/edx-enterprise/pull/69

haikuginger replied:
@pomegranited, final verdict was milliseconds.

pomegranited replied:
Sweet, thanks!

pomegranited commented on tests/test_management.py:
```diff
@@ -0,0 +1,324 @@
+    UserFactory(username='C-3PO')
+    EnterpriseCustomerFactory(
+        catalog=1,
+        name='Veridian Dynamics',
```
Serious geek creds there 😄

brittneyexline commented on integrated_channels/sap_success_factors/utils.py:
```diff
@@ -0,0 +1,96 @@
+        final_structure = {
+            'ocnCourses': self.courses
+        }
+        return json.dumps(final_structure, sort_keys=True).encode('utf-8')
```
I don't know... I will ask SAP. In terms of where that will live, I am leaning towards in the client. But this can stay for now and I can adjust when I stitch together these scripts with the transmitters/client.

haikuginger commented:
Rebased on top of master; switched to epoch seconds from milliseconds.

@brittneyexline @mattdrayer, let me know when you're good to go, and I'll go through the merge checklist.

mattdrayer commented:
MMMMMMILLLLLLLLLISSEEECCCONDDSSSS!!!!

haikuginger commented:
@brittneyexline, this is ready for merge on your final thumbs-up.
